persistence/data_models.go

```diff
@@ -187,31 +190,71 @@ type (
 		LastFailure *StateExecutionFailureJson
 	}
 
-	CompleteWaitUntilExecutionRequest struct {
+	ProcessWaitUntilExecutionRequest struct {
 		ProcessExecutionId uuid.UUID
 		StateExecutionId
 
-		Prepare        PrepareStateExecutionResponse
-		CommandRequest xdbapi.CommandRequest
-		TaskShardId    int32
+		Prepare             PrepareStateExecutionResponse
+		CommandRequest      xdbapi.CommandRequest
+		PublishToLocalQueue []xdbapi.LocalQueueMessage
+		TaskShardId         int32
 	}
 
-	CompleteWaitUntilExecutionResponse struct {
+	ProcessWaitUntilExecutionResponse struct {
 		HasNewImmediateTask bool
 	}
 
+	CompleteWaitUntilExecutionRequest struct {
+		TaskShardId        int32
+		ProcessExecutionId uuid.UUID
+		StateExecutionId
+		PreviousVersion int32
+	}
+
 	CompleteExecuteExecutionRequest struct {
 		ProcessExecutionId uuid.UUID
 		StateExecutionId
 
-		Prepare       PrepareStateExecutionResponse
-		StateDecision xdbapi.StateDecision
-		TaskShardId   int32
+		Prepare             PrepareStateExecutionResponse
+		StateDecision       xdbapi.StateDecision
+		PublishToLocalQueue []xdbapi.LocalQueueMessage
+		TaskShardId         int32
 	}
 
 	CompleteExecuteExecutionResponse struct {
 		HasNewImmediateTask bool
 	}
+
+	PublishToLocalQueueRequest struct {
+		Namespace string
+		ProcessId string
+		Messages  []xdbapi.LocalQueueMessage
+	}
+
+	PublishToLocalQueueResponse struct {
+		ProcessExecutionId  uuid.UUID
+		HasNewImmediateTask bool
+		NotExists           bool
+	}
+
+	ProcessLocalQueueMessagesRequest struct {
+		TaskShardId int32
+		// If TaskSequence is 0, skip deleting the immediate task row, and try to consume unconsumed messages for StateExecutionId
```

Comment:
Took me quite a while to understand this one, still not fully understand yet, but looks like it's for some internal implementation of the `ProcessLocalQueueMessages` persistence API. 

I would recommend to create a different internal request object for this -- put the internal request inside the persistence package so it's not leaked to outside.
```
internalProcessLocalQueueMessagesRequest struct{
       ProcessLocalQueueMessagesRequest
       isNestedRequest bool
       // below are only present when IsNestedRequest is true
       stateExecutionId   *StateExecutionId
       commandWaitingType *xdbapi.CommandWaitingType
}

Or

internalProcessLocalQueueMessagesRequest struct{
       ProcessLocalQueueMessagesRequest
       fromWaitUntilProcessing bool 
       // below are only present when IsNestedRequest is true
       stateExecutionId   *StateExecutionId
       commandWaitingType *xdbapi.CommandWaitingType
}
// seems like this is only when we are consuming when processing the waitUntil?
```
(use lower case to make them internal/private to outside)

Reply:
The meaning of `isNestedRequest ` is not clear to me. There are two cases when we will call `doProcessLocalQueueMessageTx`:

- in the client.publishToLocalQueue: do normal consume
- when a new wait_until task was executed and need to check whether it should consume the unconsumed messages: do CheckCanCompleteLocalQueueWaiting.

After either step above, the method will update state execution, update process execution, complete possible wait_until statuses.

And the TaskSequenceId is used to distingwish the two cases.

Reply:
Acutally, doing another nest level on ProcessLocalQueueMessagesRequest looks a little more cumbersome to me.

Reply:
> And the TaskSequenceId is used to distingwish the two cases.
Overloading this field is not necessary and caused confusion. 


It should be very small refactoring on what you have done:
```golang
func (p ...)ProcessNewLocalQueueMessagesTask(req ...){
    internalReq := {
        ProcessLocalQueueMessagesRequest: req
        fromWaitUntilProcessing: false
    }
    ...
    resp, err := p.doProcessNewLocalQueueMessagesTaskTx(tx, internalReq)
    ...
}

func (p ...)doProcessWaitUntilExecutionTx(tx, req ...){
   ...
   internalReq := {
       ProcessLocalQueueMessagesRequest: {...}
       fromWaitUntilProcessing: true,
       stateExecutionId: ...
       commandWaitingType: ...
   }
    resp, err := p.doProcessNewLocalQueueMessagesTaskTx(tx, internalReq)
}
```

> in the client.publishToLocalQueue: do normal consume
btw I believe you meant processing the immediate tasks in async service, because we only process in the background.

Reply:
Need to think more about how to refactor. Because I feel the ProcessLocalQueueMessagesRequest is even not necessary in the internalReq sent by `doProcessWaitUntilExecutionTx` (we can move the few used fields out from the request).

Reply:
Yeah you actually can ignore this because my latest comment is to not share this method between the two persistence APIs https://github.com/xdblab/xdb/pull/63#pullrequestreview-1697935884

we can sync on zoom for this too